File: Algorithm/Qlearning_Python/calc_ref.py
from matplotlib.pyplot import switch_backend
from helper.calc_vector_limit_angle import calc_vector, limit_angle
import math
def calc_ref(action, robot, goal, nearestObs, vmax):
    kw = 5
    dirRobot = robot[2]
    [dirGoal, disGoal] = calc_vector(robot[0:2], goal)
    radObs = nearestObs[4]
    dirObs = nearestObs[3]
    disObs = nearestObs[2]
    # % Define list of possible action
    go_ahead   = 1
    turn_left  = 2
    turn_right = 3
    pi = math.pi
    if action == go_ahead: dirRef = dirGoal
    elif action == turn_left:
        if abs(limit_angle(dirObs - dirRobot)) < math.pi/3:
            dirRef = dirObs + math.pi/2
        else:
            dirRef = dirGoal
    elif action == turn_right:
        if abs(limit_angle(dirObs - dirRobot)) < pi/3:
            dirRef = dirObs - pi/2
        else :
            dirRef = dirGoal
    vRef = vmax
    wRef = kw*limit_angle(dirRef - dirRobot)
    return vRef, wRef

File: Algorithm/QLearning_Matlab/calc_ref.m
function [vRef, wRef] = calc_ref(action, robot, goal, nearestObs)
% getAction - calculate the reference speed of the mobile robot respect to
% the action 
v = 750;
kv = 2.5;
kw = 3;
dirRobot = robot(3);
[dirGoal, disGoal] = calc_vector(robot(1:2), goal);
radObs = nearestObs(5);
disObs = nearestObs(3);
% Define list of possible action
go_ahead   = 1;
turn_left  = 2;
turn_right = 3;

switch action
    case go_ahead
        dirRef = dirGoal;
        
    case turn_left
        dirRef = dirGoal + pi/2*(radObs*2/disObs);    
        
    case turn_right
        dirRef = dirGoal - pi/2*(radObs*2/disObs);
end

if disGoal < 500
    vRef = kv*disGoal;
    dirRef = dirGoal;
else
    vRef = v;
end
wRef = kw*(dirRef - dirRobot);